tidy3d/plugins/autograd/invdes/filters.py

```diff
@@ -1,71 +1,226 @@
-    size: Union[int, Tuple[int, ...]],
+    radius: Union[float, Tuple[float, ...]] = None,
+    dl: Union[float, Tuple[float, ...]] = None,
     *,
+    size_px: Union[int, Tuple[int, ...]] = None,
     normalize: bool = True,
     padding: PaddingType = "reflect",
-):
+    filter_type: KernelType,
+) -> Callable[[NDArray], NDArray]:
     """Create a filter function based on the specified kernel type and size.
 
     Parameters
     ----------
+    radius : Union[float, Tuple[float, ...]] = None
+        The radius of the kernel. Can be a scalar or a tuple.
+    dl : Union[float, Tuple[float, ...]] = None
+        The grid spacing. Can be a scalar or a tuple.
+    size_px : Union[int, Tuple[int, ...]] = None
+        The size of the kernel in pixels for each dimension. Can be a scalar or a tuple.
+    normalize : bool = True
+        Whether to normalize the kernel so that it sums to 1.
+    padding : PaddingType = "reflect"
+        The padding mode to use.
     filter_type : KernelType
         The type of kernel to create (`circular` or `conic`).
-    size : Union[int, Tuple[int, ...]]
-        The size of the kernel in pixels for each dimension. Can be a scalar or a tuple.
-    normalize : bool, optional
-        Whether to normalize the kernel so that it sums to 1. Default is True.
-    padding : PadMode, optional
-        The padding mode to use. Default is "reflect".
 
     Returns
     -------
-    function
+    Callable[[NDArray], NDArray]
         A function that applies the created filter to an input array.
     """
-    _kernel = {}
-
-    def _filter(array):
-        original_shape = array.shape
-        squeezed_array = np.squeeze(array)
-
-        if squeezed_array.ndim not in _kernel:
-            if np.isscalar(size):
-                kernel_size = (size,) * squeezed_array.ndim
-            else:
-                kernel_size = size
-            _kernel[squeezed_array.ndim] = make_kernel(
-                kernel_type=filter_type, size=kernel_size, normalize=normalize
-            )
+    kernel_size = _get_kernel_size(radius, dl, size_px)
 
-        convolved_array = convolve(squeezed_array, _kernel[squeezed_array.ndim], padding=padding)
-        return np.reshape(convolved_array, original_shape)
+    if filter_type == "conic":
+        filter_class = ConicFilter
+    elif filter_type == "circular":
+        filter_class = CircularFilter
+    else:
+        raise ValueError(f"Unsupported filter_type: {filter_type}")
 
-    return _filter
+    filter_instance = filter_class(kernel_size=kernel_size, normalize=normalize, padding=padding)
+    return filter_instance
 
 
 make_conic_filter = partial(make_filter, filter_type="conic")
 make_conic_filter.__doc__ = """make_filter() with a default filter_type value of `conic`.
 
 See Also
 --------
-make_filter : Function to create a filter based on the specified kernel type and size.
+:func:`~filters.make_filter` : Function to create a filter based on the specified kernel type and size.
```

Comment:
what does tilde do? learning :D

Reply:
> If you prefix the content with ~, the link text will only be the last component of the target. For example, :py:meth:`~Queue.Queue.get` will refer to Queue.Queue.get but only display `get` as the link text.

from [here](https://www.sphinx-doc.org/en/master/usage/referencing.html#cross-referencing-syntax)

I just googled how to reference class methods...

Reply:
ooh, nice